### ISSUE TITLE
[HOLD] Update App.setSidebarLoaded in SidebarLinks.js to avoid rendering bug

### ISSUE THREAD

**Issue by NikkiWines** [Expensify/App#6357]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

cc: @kadiealexander 
cc: @thecodingeek as you suggested this [here](https://github.com/Expensify/App/issues/5620#issuecomment-960498522)

Put `onLayout={App.setSidebarLoaded}` in the parent `<View>` in `SidebarLinks.js` so that even if `recentReports` is an empty array,  we'll still hide the Splash Screen. 

**Note:** Because it's unclear whether or not the above modifications will actually resolve this issue the base pay for this change will be $250. If, in fact, this does resolve the persistent splash screen bug, you will be compensated for the full amount listed for https://github.com/Expensify/App/issues/5620.

## Action Performed:
Related to https://github.com/Expensify/App/issues/5620

1. Force close the app or update the app
2. Open the app

## Expected Result:
User should not be stuck on Splash screen

## Actual Result:
User is stuck in splash screen

## Workaround:
Closing and reopening app fixes the issue.

## Platform:

Where is this issue occurring?

- iOS

**Version Number:**
**Reproducible in staging?:** Yes
**Reproducible in production?:** Unknown
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
<img width="250" alt="" src="https://user-images.githubusercontent.com/16747822/142288375-af9c5cd2-d972-4f23-b57a-15fe57e8fe13.png">

**Expensify/Expensify Issue URL:** N/A
**Issue reported by:** N/A
**Slack conversation:** N/A
**Related GH Issue:** https://github.com/Expensify/App/issues/5620
**Upwork Link:** TBD

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Current assignee @kadiealexander is eligible for the External assigner, not assigning anyone new.

**Comment by kadiealexander** [comment]
https://www.upwork.com/jobs/~01a7267965ce03f7cf

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Beamanator (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kadiealexander** [comment]
@Beamanator, for context, this is a potential fix pitched by @thecodingeek for https://github.com/Expensify/App/issues/5620. We are trialling a few fixes here, and offering the full price for that issue to the user who's fix works.

@thecodingeek's proposal:

> ### Findings and Proposal
> 
> Hello everyone,
> 
> I tried to look into the causes for this issue. Below I am sharing my findings and a proposal for one reason.
> 
> As per `Expensify.js` file
> 
> ```
> if (this.getAuthToken() && this.props.initialReportDataLoaded && this.props.isSidebarLoaded) {
>     this.hideSplash();
> }
> ```
> 
> I think there are very less chance for not getting auth token, so it can most probably happen because either `initialReportDataLoaded` is false or `isSidebarLoaded` is false.
> 
> One _very interesting_ thing which I could find was, if in the file `SidebarLinks.js` in the `render` function the `recentReports` comes out to be **an empty array** in the following code,
> 
> `const {recentReports} = getSidebarOptions`
> 
> **then the Splash Screen never goes away**. Reason is when `recentReports` is empty then there are no `sections` generated and `<OptionsList` is not rendered.
> 
> Now this `<OptionsList` has the following code
> 
> ```
> <OptionsList
>     ....
>     onLayout={App.setSidebarLoaded}
> />
> ```
> 
> This method `App.setSidebarLoaded` is supposed to do the following,
> 
> `Onyx.set(ONYXKEYS.IS_SIDEBAR_LOADED, true);`
> 
> which makes `isSidebarLoaded = true`.
> 
> Now as the `OptionsList` view did not render so `onLayout` of the same is never called and `isSidebarLoaded` never becomes true. Hence it gets stuck at the Splash Screen.
> 
> An alternative (or solution) here I think can be to put `onLayout={App.setSidebarLoaded}` on the parent `<View>` **so that even if the `recentReports` is empty even then it will hide the Splash Screen and move further.** Right now it just stuck there if the recentReports is empty.
> 
> If the recentReports is actually coming out to be empty then that can be a separate issue. But I think the above handling is good to have in the code so that it never hangs at the Splash Screen even if such thing happens.
> 
> Next, the variable `initialReportDataLoaded` is set to true in the file `Report.js` in the method `fetchAllReports`
> 
> `Onyx.set(ONYXKEYS.INITIAL_REPORT_DATA_LOADED, true);`
> 
> So this scenario can happen if this `fetchAllReports` API fails, but may be that is also less likely to happen. If this is the scenario then we can do error handling there and at least move user forward with a popup or something.
> 
> If you can share the user's credentials for which this happens then I can find out the exact reason and fix even that issue as well.
> 
> @NikkiWines I would love to know what you think about this. Please check when you're available.

**Comment by sidferreira** [comment]
@NikkiWines 

> Put onLayout={App.setSidebarLoaded} in the parent <View> in SidebarLinks.js so that even if recentReports is an empty array, we'll still hide the Splash Screen.

this is what needs to be done?

A debugging fallback (not to mention https://github.com/Expensify/App/issues/6358#issuecomment-973659678 ) is to use a timeout to force the splash to hide, so we can see what's behind it. If it's all blank, or what.

And to add to the comment above, there are basically 3 items that affect the hideSplash ( https://expensify.slack.com/archives/C01GTK53T8Q/p1635878140359900?thread_ts=1635255533.107600&cid=C01GTK53T8Q )
a) Onyx migration issue
b) Sidebar loading issue
c) (part of b) some accessToken issue

The point is that even forcing some errors to happen, the app properly loads the cache (Onyx) and should work. It looks like there's an authToken expiration or losing data causing this to happen.

**Comment by NikkiWines** [comment]
> this is what needs to be done?

@sidferreira, yes that was the @thecodingeek's suggestion since his investigation [here](https://github.com/Expensify/App/issues/5620#issuecomment-960498522) revealed that if `recentReports` was `[]` we could reproduce the persistent splash screen issue. 

We will also be introducing the logic mentioned in #6358, as it's unclear if the modifications to where `App.setSidebarLoaded` lives will resolve the persistent splash screen issue.

Is your concern that implementing the changes proposed in this issue will make it harder to debug the root cause of the issue?

**Comment by sidferreira** [comment]
@NikkiWines Exactly... right now we know it happens. maybe create an investigation branch to check it out and apply the above change for now just to "keep things moving"...

**Comment by NikkiWines** [comment]
Hmm, I'm not sure I share the concern. 

If we implement this change and it fixes the root issue then great, we've identified the cause and the problem is solved. 

If it doesn't fix the solution, we've still solved a rendering issue that occurs when `recentReports` is `[]` and we now know not to further pursue this line of inquiry when investigating the persistent splash screen issue. 

I'm not opposed to a `pull to refresh` feature, but I don't really think it's necessary either. @Beamanator would be curious for your thoughts here as well.

**Comment by sidferreira** [comment]
@NikkiWines what I mean is not being able to find out what exactly was the rood cause (what makes this specific scenario happens only in a handful or people only on ios)

**Comment by Beamanator** [comment]
> @NikkiWines what I mean is not being able to find out what exactly was the rood cause (what makes this specific scenario happens only in a handful or people only on ios)

@sidferreira is your point here that "the fix for when `recentReports` is `[]`" is probably not the root cause for the specific scenario that's only happening to a handful of people only on ios?

@NikkiWines I agree with your thoughts - it seems like fixing the situation where `recentReports` can be `[]` (which is 1 way the persistent splash screen can happen) is valuable no matter what. If it doesn't fix the persistent splash screen for **everyone** (a.k.a. there's still some other super weird situation that only affects a few ios users), we'll keep pursuing solutions to that.

About `pull to refresh` - it sounded like that was proposed only so users can force the app to refresh? Is that even possible when the app is stuck on the splash screen?

**Comment by sidferreira** [comment]
@Beamanator the point is: why it comes `[]`? It should use the existing/cached content, like if it was offline. There's something else that causes the issue. Also, some `Onyx` migrations can call `hideSplash` as well, are they working?! 

The pull to refresh is a way to try to force `recentReports` to have results if we move the `App.setSidebarLoaded` to the view, but still, won't show us why `recentReports` was empty at first.

**Comment by NikkiWines** [comment]
> @Beamanator the point is: why it comes []? It should use the existing/cached content, like if it was offline. There's something else that causes the issue. Also, some Onyx migrations can call hideSplash as well, are they working?!

This is a good point @sidferreira, we'd be patching it more than providing a true fix. 

It looks like the lib owner has been commenting in the [main issue](https://github.com/Expensify/App/issues/5620) with a potential fix (or at least more insight into a potential cause). 

Let's hold off on further discussion here, if the current chats in https://github.com/Expensify/App/issues/5620 don't lead to anything we can pick back up on this?

**Comment by Beamanator** [comment]
Thanks for explaining @sidferreira 👍 

> Let's hold off on further discussion here, if the current chats in #5620 don't lead to anything we can pick back up on this?

I like it, I'll put this one on hold for now

**Comment by kadiealexander** [comment]
@Beamanator and @NikkiWines, are we still waiting on discussions from #5620? it looks like we've made good progress over there.

**Comment by NikkiWines** [comment]
@kadiealexander, yep, sorry about the lack of updates here. The lib owner has put forward a proposal that looks like it should resolve this error, in which case we probably won't need this issue. Still finalizing details there though, but I'll keep this issue updated going forward.

**Comment by kadiealexander** [comment]
As we've hired someone in the original issue who should hopefully be able to fix the issue, I'll close this for now and we can reopen if needed.
